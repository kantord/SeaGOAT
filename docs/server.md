<!-- markdownlint-disable MD046 -->
# SeaGOAT-server

The `seagoat-server` is an integral component of the Seagoat command-line tool
designed to analyze your codebase and create vector embeddings for it.

While it serves as a backend for the command-line tool, also allows you to
use it through HTTP to build your own SeaGOAT-based applications.

## Starting the server

To boot up the server for a specific repository, use:

```bash
seagoat-server start <repo_path> [--port=<custom_port>]
```

* `repo_path` - Path to your Git repository.
* `--port` - (Optional) Run the server on a specific port.

If you don't specify a custom port, a random port will be assigned to your
server. Don't worry, SeaGOAT will still be able to automatically find
the server corresponding to a specific repository.

## Developing with SeaGOAT-server

SeaGOAT-server not only serves as a backend for the SeaGOAT command-line tool
but also offers developers the capability to integrate its functions to build
custom applications.

### Retrieving server information

As SeaGOAT servers only run on one repository at a time, there is a command
provided in order to gather information about all running servers, including
how to access them through HTTP.

To get detailed information about all active SeaGOAT servers in JSON format,
you can utilize the `server-info` command:

```bash
seagoat-server server-info
```

You will receive a response similar to this:

```json
{
    "version": "0.5.3",
    "globalCache": "/home/myuser/.cache/seagoat",
    "globalConfigFile": "/home/myuser/.config/seagoat/config.yml",
    "servers": {
        "/path/to/repository/1": {
            "cacheLocation": {
              "chroma": "/home/myuser/.cache/seagoat/bfe8133b9e871ea1c8498a0"
            },
            "isRunning": true,
            "host": "127.0.0.1",
            "port": "8080",
            "address": "http://127.0.0.1:8080"
        },
        "/path/to/repository/2": {
            "cacheLocation": {
              "chroma": "/home/myuser/.cache/seagoat/fbee39c83bd47a75e2f839"
            },
            "isRunning": false,
            "host": "127.0.0.1",
            "port": "8081",
            "address": "http://127.0.0.1:8081"
        }
    }
}
```

In this output, you can also see information about where databases/caches
related to your projects are stored. `globalCache` is the parent folder of
all the cache directories, and withing each server, you can find an attribute
called `cacheLocation` which contains the path to the cache directory for
each different type of cache associated with that project.

If you want to create a configuration file, you can see the path for it
in the `globalConfigFile` attribute. This depends on your operating system.
You can also create a configuration file for your project. See
[the configuration documentation](configuration.md) for more information.

### Making queries using the API

If you want to build an application using SeaGOAT-server, first you need to
figure out the address of the server you want to connect to.

Once you have the address, you can start making queries to it. For instance,
this is how you'd make a query using `curl` to the server running on
`http://localhost:32835`:

```bash
curl 'http://localhost:32835/query/example'
```

You will receive a response similar to this one:

```json
{
  "results": [
    {
      "path": "tests/conftest.py",
      "fullPath": "/home/user/repos/SeaGOAT/tests/conftest.py",
      "score": 0.6,
      "blocks": [
        {
          "score": 0.21,
          "lines": [
            {
              "score": 0.21,
              "line": 100,
              "lineText": "def very_relevant_function():",
              "resultTypes": [
                "result"
              ]
            }
          ],
          "lineTypeCount": {
            "result": 1
          }
        },
        {
          "score": 0.6,
          "lines": [
            {
              "score": 0.6,
              "line": 489,
              "lineText": " contents=(\"hello()\\n\" * (i % 50)),",
              "resultTypes": [
                "result"
              ]
            },
            {
              "score": 0.84,
              "line": 490,
              "lineText": "     return foo * bar",
              "resultTypes": [
                "result"
              ]
            }
          ],
          "lineTypeCount": {
            "result": 1
          }
        }
      ]
    },
    {
      "path": "tests/test_cli.py",
      "fullPath": "/home/user/repos/SeaGOAT/tests/test_cli.py",
      "score": 0.87,
      "blocks": [... etc ... ]
    },
    ... etc ...
  ],
  "version": "0.26.0"
}
```

#### Understanding the response

The response contains the following information:

* `version` - This is the version of SeaGOAT being used.
* `results` - This is an array containing your results.

Each result inside results has the following data:

* `path` - The (relative) path of the file within the repository.
* `fullPath` - The absolute path to the file in your filesystem.
* `score` - A number indicating how relevant a result is, smaller is better.
* `blocks` - An array of relevant code blocks from this file.

Within each block you will find:

* `lines` - An array of line objects containing:
  * `score` - Relevance score for this line. See `score` above.
  * `line` - The line number in the file where the result was found.
  * `lineText` - The actual text content of that line.
  * `resultTypes` - An array indicating all types of result on this line:
    * `"result"` Means that the line is directly relevant to the query.
    * `"context"` Means that the line was added as a context line.
    * `"bridge"` It is a type of context line that is added within a block
    rather than around it. It is used to merge code blocks when they are
    nearly touching.
* `lineTypeCount` - An object containing a count of all line types within
  the code block. See `resultTypes` for more.
* `score` - A score for the code block overall.
