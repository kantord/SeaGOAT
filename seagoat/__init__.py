__version__ = "0.49.4"
