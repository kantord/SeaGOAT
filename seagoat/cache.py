import hashlib
import os
import pickle
from pathlib import Path
from typing import Generic
from typing import TypeVar

import appdirs

T = TypeVar("T")

CACHE_FORMAT_VERSION = 15


class Cache(Generic[T]):
    def __init__(self, cache_name: str, path: Path, initial_value: T):
        self._path = path
        self.data = initial_value
        self._cache_name = cache_name

    def load(self):
        try:
            with open(self._get_cache_file(), "rb") as cache_file:
                self.data = pickle.load(cache_file)
        except (FileNotFoundError, pickle.UnpicklingError, EOFError):
            pass

    def persist(self):
        with open(self._get_cache_file(), "wb") as cache_file:
            pickle.dump(self.data, cache_file)

    def _get_cache_file(self):
        return self.get_cache_folder() / self._cache_name

    def get_cache_folder(self):
        cache_folder = self._get_cache_root() / self._get_project_hash()
        cache_folder.mkdir(parents=True, exist_ok=True)

        return cache_folder

    def _get_cache_root(self):
        if "RUNNER_TEMP" in os.environ:
            return Path(os.environ["RUNNER_TEMP"])

        return Path(
            appdirs.user_cache_dir(
                "seagoat-pytest" if "PYTEST_CURRENT_TEST" in os.environ else "seagoat"
            )
        )

    def _get_project_hash(self):
        normalized_path = Path(self._path).expanduser().resolve()
        text = f"""
        Cache version: {CACHE_FORMAT_VERSION}
        Normalized path: {normalized_path}
        """

        return hashlib.sha256(text.encode()).hexdigest()
