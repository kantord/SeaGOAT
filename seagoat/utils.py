import json
import os

import appdirs


def get_server_info_file(repo_path):
    repo_id = os.path.normpath(repo_path).replace(os.sep, "_")
    user_cache_dir = appdirs.user_cache_dir("seagoat-servers")
    os.makedirs(user_cache_dir, exist_ok=True)
    return os.path.join(user_cache_dir, f"{repo_id}.json")


def load_server_info(server_info_file):
    with open(server_info_file, "r", encoding="utf-8") as file:
        server_info = json.load(file)
    host = server_info["host"]
    port = server_info["port"]
    pid = server_info.get("pid")
    server_address = f"http://{host}:{port}"
    return host, port, pid, server_address
