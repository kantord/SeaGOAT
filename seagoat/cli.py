import os
from functools import cache

import click
import requests
from pygments import highlight
from pygments.formatters import TerminalFormatter
from pygments.lexers import get_lexer_for_filename
from pygments.lexers.javascript import JavascriptLexer
from pygments.lexers.javascript import TypeScriptLexer

from seagoat.server import get_server_info_file
from seagoat.server import load_server_info


def query_server(query, server_address):
    response = requests.get(f"{server_address}/query/{query}")
    try:
        response.raise_for_status()
    except requests.exceptions.HTTPError as err:
        print("HTTP error occurred:")
        print(f"Response code: {response.status_code}")
        print(f"Error: {err}")
        print(f"Response body: {response.text}")
        raise
    return response.json()["results"]


@cache
def get_highlighted_lines(file_name: str):
    with open(file_name, "r", encoding="utf-8") as source_code_file:
        code = source_code_file.read()

    if file_name.endswith(".md"):
        return code.splitlines()

    if file_name.endswith(".jsx"):
        lexer = JavascriptLexer()
    elif file_name.endswith(".tsx"):
        lexer = TypeScriptLexer()
    else:
        lexer = get_lexer_for_filename(file_name)

    result = highlight(code, lexer, TerminalFormatter())

    return result.splitlines()


def print_result_line(result, line, color_enabled):
    if color_enabled:
        highlighted_lines = get_highlighted_lines(str(result["full_path"]))
        print(
            f"{result['path']}:{click.style(str(line), bold=True)}:{highlighted_lines[line - 1]}"
        )
    else:
        for line_content in result["lines"]:
            if line_content["line"] == line:
                print(f"{result['path']}:{line}:{line_content['line_text']}")
                break


@click.command()
@click.argument("query")
@click.argument("repo_path", required=False, default=os.getcwd())
@click.option(
    "--no-color",
    is_flag=True,
    help="Disable formatting. Automatically enabled when part of a bash pipeline.",
)
def seagoat(query, repo_path, no_color):
    """
    Query your codebase for your QUERY in the Git repository REPO_PATH.

    Your query can either be text that you expect to find in a file,
    or a description of what you are looking for.

    When REPO_PATH is not specified, the current working directory is
    assumed to be the repository path.
    """
    _, __, ___, server_address = load_server_info(get_server_info_file(repo_path))
    results = query_server(query, server_address)

    color_enabled = os.isatty(0) and not no_color
    for result in results:
        for result_line in result.get("lines", []):
            print_result_line(result, result_line["line"], color_enabled)


if __name__ == "__main__":
    # pylint: disable-next=no-value-for-parameter
    seagoat()
