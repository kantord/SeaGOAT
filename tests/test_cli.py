import os
import re
import shutil
import tempfile
from typing import List

import orjson
import pytest
import requests
from click.testing import CliRunner

from seagoat import __version__
from seagoat.cli import query_server, seagoat, warn_if_update_available
from seagoat.utils.cli_display import is_bat_installed
from seagoat.utils.server import update_server_info
from tests.conftest import MagicMock


@pytest.fixture
def real_bat():
    return


@pytest.fixture
def lots_of_fake_results(mocker):
    mock_results = {
        "results": [
            {
                "blocks": [
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 9,
                                "lineText": "* feat: only display full code blocks in result ([`e2767f9`](https://github.com/kantord/SeaGOAT/commit/e2767f98ece3023e01ec4a6d95cd14701b11f842))",
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 24,
                                "lineText": "* refactor: group continuous lines into blocks ([`6a15673`](https://github.com/kantord/SeaGOAT/commit/6a15673e27bf906f8aca4e7eb000a8e24ba56acf))",
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                ],
                "fullPath": "/home/kantord/repos/SeaGOAT/CHANGELOG.md",
                "path": "CHANGELOG.md",
            },
            {
                "blocks": [
                    {
                        "lineTypeCount": {"result": 2},
                        "lines": [
                            {
                                "line": 240,
                                "lineText": '                    print("Server responded with a 500 error:")',
                                "resultTypes": ["result"],
                            },
                            {
                                "line": 241,
                                "lineText": "                    print(response.text)",
                                "resultTypes": ["result"],
                            },
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 2},
                        "lines": [
                            {
                                "line": 283,
                                "lineText": "    def _mock_results(results_template):",
                                "resultTypes": ["result"],
                            },
                            {
                                "line": 284,
                                "lineText": "        results = []",
                                "resultTypes": ["result"],
                            },
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 2},
                        "lines": [
                            {
                                "line": 287,
                                "lineText": "        for filename, lines in results_template:",
                                "resultTypes": ["result"],
                            },
                            {
                                "line": 288,
                                "lineText": "            for i, line_text in enumerate(lines):",
                                "resultTypes": ["result"],
                            },
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 291,
                                "lineText": "        for filename, lines in fake_files.items():",
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 295,
                                "lineText": "                lines_sorted_by_line_number = sorted(lines.items(), key=lambda x: x[0])",
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 297,
                                "lineText": "                    [line for _, line in lines_sorted_by_line_number]",
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 2},
                        "lines": [
                            {
                                "line": 301,
                                "lineText": "        for filename, lines in results_template:",
                                "resultTypes": ["result"],
                            },
                            {
                                "line": 302,
                                "lineText": "            result = {",
                                "resultTypes": ["result"],
                            },
                        ],
                    },
                ],
                "fullPath": "/home/kantord/repos/SeaGOAT/tests/conftest.py",
                "path": "tests/conftest.py",
            },
            {
                "blocks": [
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 43,
                                "lineText": '        results = current_app.extensions["task_queue"].enqueue_high_prio(',
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 51,
                                "lineText": "        for result in results:",
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 53,
                                "lineText": "                result.add_context_lines(-int(context_above))",
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                    {
                        "lineTypeCount": {"result": 1},
                        "lines": [
                            {
                                "line": 56,
                                "lineText": '            "results": [result.to_json(query) for result in results],',
                                "resultTypes": ["result"],
                            }
                        ],
                    },
                ],
                "fullPath": "/home/kantord/repos/SeaGOAT/seagoat/server.py",
                "path": "seagoat/server.py",
            },
        ],
        "version": "0.18.0",
    }

    mock_response = MagicMock()
    mock_response.text = orjson.dumps(mock_results)
    mocker.patch("requests.post", return_value=mock_response)

    return mock_results


@pytest.fixture(name="get_request_args_from_cli_call")
def get_request_args_from_cli_call_(mock_server_factory, mocker, runner, repo):
    def _run_cli(cli_args: List[str]):
        mock_server_factory(
            [
                ["hello.txt", ["foo", "bar", "baz"]],
                ["foobar.py", ["def hello():", "    print('world')"]],
                ["foobar2.py", ["def hola():", "    print('mundo')"]],
                ["foobar3.py", ["def bonjour():", "    print('monde')"]],
                ["foobar.fake", ["fn hello()", "    echo('world')"]],
                ["foobar2.fake", ["fn hola()", "    echo('mundo')"]],
                ["foobar3.fake", ["fn bonjour()", "    echo('monde')"]],
            ],
            manually_mock_request=True,
        )

        request_args = {}

        def fake_requests(*args, **kwargs):
            mock_response = mocker.Mock()
            mock_response.text = '{"results": []}'

            request_args["url"] = args[0]
            request_args["json"] = kwargs.get("json", {})

            return mock_response

        mocked_requests_post = mocker.patch(
            "seagoat.cli.requests.post", side_effect=fake_requests
        )
        query = "JavaScript"
        result = runner.invoke(
            seagoat,
            [query, repo.working_dir, "--no-color", *cli_args],
        )
        assert result.exit_code == 0
        mocked_requests_post.assert_called_once()

        return request_args

    return _run_cli


@pytest.fixture(name="mock_response")
def mock_response_(mocker):
    def _mock_response(json_data, status_code=200):
        mock_resp = mocker.Mock(spec=requests.Response)
        mock_resp.status_code = status_code
        mock_resp.text = orjson.dumps(json_data)
        return mock_resp

    return _mock_response


@pytest.fixture
def mock_accuracy_warning(mocker):
    def _noop(*args, **kwargs):
        pass

    mocker.patch("seagoat.cli.display_accuracy_warning", _noop)


@pytest.fixture
def mock_query_server(mocker):
    def _mocked_query_server(*args, **kwargs):
        return []

    mocker.patch("seagoat.cli.query_server", _mocked_query_server)


@pytest.mark.usefixtures("mock_query_server")
@pytest.mark.parametrize("accuracy_percentage", [50, 75, 99, 100])
def test_seagoat_warns_on_incomplete_accuracy(
    accuracy_percentage,
    runner_with_error,
    mocker,
    mock_response,
):
    mock_status_response = {
        "stats": {
            "chunks": {"analyzed": 100, "unanalyzed": 0},
            "queue": {"size": 0},
            "accuracy": {"percentage": accuracy_percentage},
        },
        "version": __version__,
    }

    mocker.patch("requests.get", return_value=mock_response(mock_status_response))
    mocker.patch("seagoat.cli.get_server_info", return_value={"address": ""})

    query = "some random query"
    result = runner_with_error.invoke(seagoat, [query, "--no-color"])

    if accuracy_percentage < 100:
        assert (
            "Warning: SeaGOAT is still analyzing your repository. "
            + f"The results displayed have an estimated accuracy of {accuracy_percentage}%"
            in result.stderr
        )
    else:
        assert "Warning" not in result.stderr


@pytest.mark.usefixtures("mock_accuracy_warning")
@pytest.mark.parametrize("max_length", [1, 2, 10])
def test_forwards_limit_clue_to_server(max_length, get_request_args_from_cli_call):
    request_args = get_request_args_from_cli_call(["--max-results", str(max_length)])
    assert request_args["json"]["limitClue"] == max_length


@pytest.mark.usefixtures("server", "mock_accuracy_warning")
def test_integration_test_without_color(snapshot, repo, mocker, runner, temporary_cd):
    mocker.patch("os.isatty", return_value=True)
    query = "JavaScript"
    with temporary_cd(repo.working_dir):
        result = runner.invoke(seagoat, [query, ".", "--no-color"])

    assert str(result.output) == snapshot
    assert result.exit_code == 0


@pytest.mark.usefixtures("server", "mock_accuracy_warning")
def test_integration_test_vimgrep_mode(snapshot, repo, mocker, runner, temporary_cd):
    mocker.patch("os.isatty", return_value=True)
    query = "JavaScript"
    with temporary_cd(repo.working_dir):
        result = runner.invoke(seagoat, [query, ".", "--vimgrep"])

    assert str(result.output) == snapshot
    assert result.exit_code == 0


@pytest.mark.usefixtures("mock_accuracy_warning")
@pytest.mark.parametrize(
    "max_length, command_option, expected_length",
    [
        (20, "-l", 20),
        (0, "-l", 3),
        (1, "--max-results", 3),
        (3, "--max-results", 3),
        (4, "--max-results", 5),
        (6, "-l", 7),
        (2, "-l", 3),
    ],
)
def test_limit_output_length(
    repo,
    max_length,
    command_option,
    mock_server_factory,
    runner,
    expected_length,
):
    mock_server_factory(
        [
            ["hello.txt", ["foo", "bar", "baz"]],
            ["foobar.py", ["def hello():", "    print('world')"]],
            ["foobar2.py", ["def hola():", "    print('mundo')"]],
            ["foobar3.py", ["def bonjour():", "    print('monde')"]],
            ["foobar.fake", ["fn hello()", "    echo('world')"]],
            ["foobar2.fake", ["fn hola()", "    echo('mundo')"]],
            ["foobar3.fake", ["fn bonjour()", "    echo('monde')"]],
        ]
    )
    query = "JavaScript"
    result = runner.invoke(
        seagoat,
        [
            query,
            repo.working_dir,
            "--no-color",
            command_option,
            str(max_length),
        ],
    )

    assert len(result.output.splitlines()) == min(expected_length, 15)
    assert result.output.splitlines() == result.output.splitlines()[:expected_length]
    assert result.exit_code == 0


def test_version_option(runner):
    result = runner.invoke(seagoat, ["--version"])

    assert result.exit_code == 0
    assert result.output.strip() == f"seagoat, version {__version__}"


@pytest.mark.parametrize(
    "repo_path",
    [
        ("/path/to/repo1"),
        ("/another/path/to/repo2"),
    ],
)
def test_server_is_not_running_error(mocker, repo_path, snapshot):
    update_server_info(
        repo_path,
        {
            "host": "localhost",
            "port": 345435,
            "pid": 234234,
            "address": "http://example.com",
        },
    )
    runner = CliRunner()

    mocker.patch("os.isatty", return_value=True)
    query = "JavaScript"
    result = runner.invoke(seagoat, [query, repo_path])

    assert result.exit_code == 3
    assert str(result.output) == snapshot


def test_documentation_present(runner):
    result = runner.invoke(seagoat, ["--help"])
    assert result.exit_code == 0
    assert "Query your codebase for your QUERY" in result.output


@pytest.mark.usefixtures("mock_accuracy_warning")
@pytest.mark.parametrize("context_above", [1, 2, 10])
def test_forwards_context_above_to_server(
    context_above, get_request_args_from_cli_call
):
    request_args1 = get_request_args_from_cli_call(
        ["--context-above", str(context_above)]
    )
    assert request_args1["json"]["contextAbove"] == context_above

    request_args2 = get_request_args_from_cli_call([f"-B{context_above}"])
    assert request_args2["json"]["contextAbove"] == context_above


@pytest.mark.usefixtures("mock_accuracy_warning")
@pytest.mark.parametrize("context_below", [1, 2, 10])
def test_forwards_context_below_to_server(
    context_below, get_request_args_from_cli_call
):
    request_args1 = get_request_args_from_cli_call(
        ["--context-below", str(context_below)]
    )
    assert request_args1["json"]["contextBelow"] == context_below

    request_args2 = get_request_args_from_cli_call([f"-A{context_below}"])
    assert request_args2["json"]["contextBelow"] == context_below


@pytest.mark.usefixtures("mock_accuracy_warning")
@pytest.mark.parametrize("context", [0, 1, 2, 10])
def test_forwards_context_to_server(context, get_request_args_from_cli_call):
    request_args1 = get_request_args_from_cli_call(["--context", str(context)])
    assert request_args1["json"]["contextAbove"] == context
    assert request_args1["json"]["contextBelow"] == context

    request_args2 = get_request_args_from_cli_call([f"-C{context}"])
    assert request_args2["json"]["contextAbove"] == context
    assert request_args2["json"]["contextBelow"] == context


@pytest.mark.usefixtures("mock_accuracy_warning")
def test_limit_does_not_apply_to_context_lines(repo, mock_server_factory, runner):
    mock_server_factory(
        [
            ["hello.txt", ["foo", "a context line", "baz"]],
            ["foobar.py", ["def hello():", "    print('world')"]],
            [
                "foobar2.py",
                [
                    "def hola():",
                    "    print('mundo') # I am a context line",
                    "# screaming: context!",
                ],
            ],
            ["foobar3.py", ["def bonjour():", "    print('monde')"]],
            [
                "foobar.fake",
                [
                    "fn hello()",
                    "    echo('world') // I am also a context line",
                ],
            ],
            ["foobar2.fake", ["fn hola()", "    echo('mundo')"]],
            [
                "foobar3.fake",
                [
                    "fn bonjour()",
                    "    echo('monde') /* I am too a context line */",
                ],
            ],
        ]
    )
    query = "JavaScript"
    result = runner.invoke(
        seagoat,
        [query, repo.working_dir, "--no-color", "-l5", "--context=3"],
    )

    assert len(result.output.splitlines()) > 5
    assert result.exit_code == 0


@pytest.mark.usefixtures("mock_accuracy_warning")
def test_context_lines_at_the_end_are_included(repo, mock_server_factory, runner):
    mock_server_factory(
        [
            [
                "hello.py",
                [
                    "foo()",
                    "# a context line",
                    "baz()",
                    "# another context line",
                    "# 3rd context line",
                ],
            ],
        ]
    )
    query = "JavaScript"
    result = runner.invoke(
        seagoat,
        [query, repo.working_dir, "--no-color", "-l2", "--context=3"],
    )

    assert len(result.output.splitlines()) == 5
    assert result.exit_code == 0


@pytest.mark.usefixtures("mock_accuracy_warning")
def test_context_lines_are_not_included_from_other_files_when_limit_exceeded(
    repo, mock_server_factory, runner
):
    mock_server_factory(
        [
            [
                "hello.py",
                [
                    "foo()",
                    "# a context line",
                    "baz()",
                    "# another context line",
                    "# 3rd context line",
                ],
            ],
            [
                "hello2.py",
                [
                    "# a context line",
                    "baz()",
                    "# another context line",
                    "# 3rd context line",
                ],
            ],
        ]
    )
    query = "JavaScript"
    result = runner.invoke(
        seagoat,
        [query, repo.working_dir, "--no-color", "-l2", "--context=3"],
    )

    assert len(result.output.splitlines()) == 5
    assert result.exit_code == 0


@pytest.mark.usefixtures("mock_accuracy_warning")
@pytest.mark.parametrize("args", [(), ("--reverse",)], ids=["default", "reversed"])
def test_reverse_ordering(
    repo, mock_server_factory, runner, snapshot, args: tuple[str, ...]
):
    mock_server_factory(
        [
            ["hello.txt", ["foo", "a context line", "baz"]],
            ["foobar.py", ["def hello():", "    print('world')"]],
        ]
    )
    query = "JavaScript"
    result = runner.invoke(
        seagoat,
        [query, repo.working_dir, "--no-color", *args],
    )

    assert str(result.output) == snapshot
    assert result.exit_code == 0


@pytest.mark.parametrize(
    "error_message, error_code",
    [("File Not Found on Server", 500), ("Database Connection Failed", 503)],
)
def test_server_error_handling(
    repo,
    mock_server_error_factory,
    runner_with_error,
    error_message,
    error_code,
):
    mock_server_error_factory(error_message, error_code)

    query = "JavaScript"
    result = runner_with_error.invoke(
        seagoat,
        [query, repo.working_dir, "--no-color", "-l2", "--context=3"],
    )

    assert result.exit_code == 4
    assert error_message in result.stderr


def test_bat_installed(mocker):
    mock_run = mocker.patch("seagoat.utils.cli_display.subprocess.run")
    mock_run.return_value.returncode = 0
    assert is_bat_installed() is True


def test_bat_not_installed_1(mocker):
    mocker.patch(
        "seagoat.utils.cli_display.subprocess.run",
        side_effect=FileNotFoundError,
    )
    assert is_bat_installed() is False


def test_bat_not_installed_2(mocker):
    mock_run = mocker.patch("seagoat.utils.cli_display.subprocess.run")
    mock_run.return_value.returncode = 1
    assert is_bat_installed() is False


@pytest.mark.parametrize("newer_version", ["4243.3424.2", "99992.0.0"])
def test_warn_if_update_available_shows_warning(
    mocker, capsys, mock_response, newer_version
):
    mock_version_response = {"info": {"version": newer_version}}

    mocker.patch("requests.get", return_value=mock_response(mock_version_response))

    warn_if_update_available()
    captured = capsys.readouterr()

    assert (
        f"Warning: An updated version {newer_version} of SeaGOAT is available. You have {__version__}."
        in captured.err
    )


def test_warn_if_update_available_no_warning(mocker, capsys, mock_response):
    same_version = __version__
    mock_version_response = {"info": {"version": same_version}}

    mocker.patch("requests.get", return_value=mock_response(mock_version_response))

    warn_if_update_available()
    captured = capsys.readouterr()

    assert "Warning" not in captured.err


@pytest.mark.usefixtures(
    "server",
    "mock_accuracy_warning",
    "bat_available",
    "mock_warn_if_update_available",
)
def test_integration_test_no_results(snapshot, repo, mocker, runner):
    mocker.patch("os.isatty", return_value=True)
    mocker.patch("seagoat.utils.cli_display.display_blocks_with_bat", return_value=True)
    query = "a_string_we_are_sure_does_not_exist_in_any_file_12345"
    result = runner.invoke(seagoat, [query, repo.working_dir])
    assert str(result.output) == snapshot
    assert result.exit_code == 0


@pytest.mark.usefixtures("mock_accuracy_warning")
@pytest.mark.parametrize(
    "remote_host",
    ["http://example.com/potato", "https://ejemplo.es/nose/seagoat"],
)
def test_configure_remote_server_address(
    remote_host, get_request_args_from_cli_call, create_config_file
):
    create_config_file({"client": {"host": remote_host}})
    request_args = get_request_args_from_cli_call([])
    assert request_args["url"].startswith(remote_host)


@pytest.mark.usefixtures("mock_accuracy_warning", "bat_available")
def test_connecting_to_remote_server(
    repo,
    mocker,
    runner,
    temporary_cd,
    server,
    create_config_file,
    bat_calls,
    snapshot,
):
    """
    When the user requests data from a remote server,
    the display logic should not rely on full path
    as the full path might be different on the remote server
    than locally

    Also the same files might not always exist.

    This test also tests that results are correctly displayed with bat
    """

    create_config_file({"client": {"host": server}})
    mocker.patch("os.isatty", return_value=True)
    query = "JavaScript"

    def query_server_side_effect(*args, **kwargs):
        results = query_server(*args, **kwargs)

        results = [{**result, "fullPath": "/non/existent/path/"} for result in results]

        return results

    mocker.patch("seagoat.cli.query_server", side_effect=query_server_side_effect)

    with tempfile.TemporaryDirectory() as tmpdirname:
        shutil.copytree(
            repo.working_dir,
            os.path.join(tmpdirname, "repo_copy"),
            dirs_exist_ok=True,
        )
        repo.add_file_change_commit(
            file_name="example_should_not_exist_in_copy.txt",
            contents="""JavaScript is an amazing programming language""",
            author=repo.actors["John Doe"],
            commit_message="Add fruits data",
        )
        with temporary_cd(os.path.join(tmpdirname, "repo_copy")):
            result = runner.invoke(seagoat, [query, "."])

    assert result.exit_code == 0

    for bat_call in bat_calls:
        assert "example_should_not_exist_in_copy.txt" not in str(bat_call)

    assert [
        re.sub(r" .*repo_copy[/|\\]", " normalized_repo_path/", call)
        for call in bat_calls
    ] == snapshot

    assert len(bat_calls) == 2


def test_server_does_not_exist_error(runner_with_error, mocker, repo):
    mocker.patch(
        "seagoat.cli.requests.get",
        side_effect=requests.exceptions.ConnectionError,
    )
    query = "JavaScript"
    result = runner_with_error.invoke(seagoat, [query, repo.working_dir])

    assert result.exit_code == 3
    assert "The SeaGOAT server is not running" in result.output


@pytest.mark.usefixtures("mock_query_server")
def test_no_network_to_update(runner_with_error, mocker, mock_response, repo):
    mocked_response = {
        "stats": {
            "chunks": {"analyzed": 100, "unanalyzed": 0},
            "queue": {"size": 0},
            "accuracy": {"percentage": 100},
        },
        "version": __version__,
    }

    mocker.patch("requests.get", return_value=mock_response(mocked_response))
    mocker.patch("seagoat.cli.get_server_info", return_value={"address": ""})
    mocker.patch(
        "seagoat.cli.warn_if_update_available",
        side_effect=requests.exceptions.ConnectionError,
    )
    result = runner_with_error.invoke(seagoat, ["search", repo.working_dir])

    assert result.exit_code == 0
    assert "Could not check for updates" in result.output


@pytest.mark.usefixtures("server", "mock_accuracy_warning")
def test_does_not_crash_with_slash_in_request(
    snapshot, repo, mocker, runner, temporary_cd
):
    mocker.patch("os.isatty", return_value=True)
    query = "JavaScript/TypeScript"

    with temporary_cd(repo.working_dir):
        result = runner.invoke(seagoat, [query, ".", "--no-color"])

    assert result.exit_code == 0
