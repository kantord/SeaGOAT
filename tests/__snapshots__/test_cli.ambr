# serializer version: 1
# name: test_connecting_to_remote_server
  list([
    'bat normalized_repo_path/file4.js --file-name file4.js --paging never --line-range 1:1',
    'bat normalized_repo_path/file4.md --file-name file4.md --paging never --line-range 1:1',
  ])
# ---
# name: test_integration_test_no_results
  ''
# ---
# name: test_integration_test_vimgrep_mode
  '''
  file4.js:1:0:// This is a second updated example JavaScript file
  file4.md:1:0:// This is a second updated example JavaScript file
  
  '''
# ---
# name: test_integration_test_without_color
  '''
  file4.js:1:// This is a second updated example JavaScript file
  file4.md:1:// This is a second updated example JavaScript file
  
  '''
# ---
# name: test_reverse_ordering[default]
  '''
  hello.txt:1:foo
  hello.txt:2:a context line
  hello.txt:3:baz
  foobar.py:1:def hello():
  foobar.py:2:    print('world')
  
  '''
# ---
# name: test_reverse_ordering[reversed]
  '''
  foobar.py:1:def hello():
  foobar.py:2:    print('world')
  hello.txt:1:foo
  hello.txt:2:a context line
  hello.txt:3:baz
  
  '''
# ---
# name: test_server_is_not_running_error[/another/path/to/repo2]
  '''
  The SeaGOAT server is not running. Please start the server using the following command:
  
  seagoat-server start /another/path/to/repo2
  
  
  '''
# ---
# name: test_server_is_not_running_error[/path/to/repo1]
  '''
  The SeaGOAT server is not running. Please start the server using the following command:
  
  seagoat-server start /path/to/repo1
  
  
  '''
# ---
